src/main/java/com/snowflake/kafka/connector/internal/streaming/StreamingClientProvider.java

```diff
@@ -38,18 +40,29 @@
  * node with equal {@link StreamingClientProperties} will use the same client
  */
 public class StreamingClientProvider {
-  private static class StreamingClientProviderSingleton {
-    private static final StreamingClientProvider streamingClientProvider =
-        new StreamingClientProvider();
-  }
+  private static StreamingClientProvider streamingClientProvider = null;
+
+  private static Supplier<StreamingClientHandler> clientHandlerSupplier =
+      DirectStreamingClientHandler::new;
 
   /**
    * Gets the current streaming provider
    *
    * @return The streaming client provider
    */
   public static StreamingClientProvider getStreamingClientProviderInstance() {
-    return StreamingClientProviderSingleton.streamingClientProvider;
+    if (streamingClientProvider == null) {
+      streamingClientProvider = new StreamingClientProvider(clientHandlerSupplier.get());
+    }
+
+    return streamingClientProvider;
+  }
+
+  public static void overrideStreamingClientHandler(StreamingClientHandler streamingClientHandler) {
```

Comment:
Let's add a doc explaining it's used for tests only.

Reply:
done, added annotation as well